### PR TITLE
Improved unstable channel error handling

### PR DESCRIPTION
**Pull request by patrick-wright**
… handling

1) case where required artifactory env vars are not set when using unstable channel
2) case where artifactory server can not be resolved
3) case where artifactory server returns 401 for bad credentials
4) case where user adds trailing slash to ENDPOINT

@sersut @schisamo first draft  going to refine more tomorrow and add the case where we get no results from querying artifactory or omnitruck

### CODE REVIEW

File: lib/mixlib/install/backend/artifactory.rb [hunk 1]
```diff
@@ -25,6 +25,11 @@ module Mixlib
   class Install
     class Backend
       class Artifactory
+        class ConnectionError < StandardError; end
+        class AuthenticationError < StandardError; end
+
+        ENDPOINT = "http://artifactory.chef.co"
```
Comment (schisamo) [inline lib/mixlib/install/backend/artifactory.rb]:
```ruby
ENDPOINT = "http://artifactory.chef.co".freeze
```

File: lib/mixlib/install/options.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ module Mixlib
   class Install
     class Options
       class InvalidOptions < ArgumentError; end
+      class EnvironmentVariablesNotFound < StandardError; end
```
Comment (sersut) [inline lib/mixlib/install/options.rb]: Let's rename this `ArtifactoryCredentialsNotFound`. I think that is a bit more clear in terms of what the error means.

File: mixlib-install.gemspec [hunk 3]
```diff
@@ -24,4 +24,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "rspec", "~> 3.3"
   spec.add_development_dependency "pry"
+  spec.add_development_dependency "climate_control"
```
Comment (sersut) [inline mixlib-install.gemspec]: Looks like this gem brings in ActiveSupport [as a dependency](https://github.com/thoughtbot/climate_control/blob/master/climate_control.gemspec#L20) to make things thread-safe. Do we really need that baggage here? 

We can bring in some simplified version of [these](https://github.com/thoughtbot/climate_control/blob/master/lib/climate_control/modifier.rb) functions into our spec_helper and we should be good to go.

**Comment by sersut** [comment]
Super awesomeness @patrick-wright. Other than my comment about the climate_control gem I'm :+1:

File: lib/mixlib/install/options.rb [hunk 4]
```diff
@@ -132,6 +134,13 @@ def validate_shell_type
 Must be one of: #{SUPPORTED_SHELL_TYPES.join(", ")}"
         end
       end
+
+      def validate_unstable_channel
```
Comment (schisamo) [inline lib/mixlib/install/options.rb]: Ruby methods that raise should end in a bang `!`.

**Comment by sersut** [comment]
:+1:

**Comment by sersut** [comment]
@patrick-wright code :+1: 

Unstable tests seem to be failing with this:

```
Failures:

  1) Mixlib::Install::Backend for chef for unstable when p, pv and m are present for mac with an integration product version behaves like the right artifact info gives the right url artifact info
     Failure/Error: expect(info.url).to match expected_info[:url]
       expected "http://localhost:8080/artifactory/omnibus-unstable-local/com/getchef/chef/12.5.1+20151210002019/mac_os_x/10.9/chef-12.5.1+20151210002019-1.dmg" to match "http://artifactory.chef.co/omnibus-unstable-local/com/getchef/chef/12.5.1+20151210002019/mac_os_x/10.9/chef-12.5.1+20151210002019-1.dmg"
     Shared Example Group: "the right artifact info" called from ./spec/mixlib/install/backend_spec.rb:352
     # ./spec/mixlib/install/backend_spec.rb:50:in `block (3 levels) in <top (required)>'

  2) Mixlib::Install::Backend for chef for unstable when p, pv and m are present for windows behaves like the right artifact info gives the right url artifact info
     Failure/Error: expect(info.url).to match expected_info[:url]
       expected "http://localhost:8080/artifactory/omnibus-unstable-local/com/getchef/chef/12.5.1+20151210002019/windows/2012r2/chef-client-12.5.1+20151210002019-1-x86.msi" to match "http://artifactory.chef.co/omnibus-unstable-local/com/getchef/chef/12.5.1+20151210002019/windows/2012r2/chef-client-12.5.1+20151210002019-1-x86.msi"
     Shared Example Group: "the right artifact info" called from ./spec/mixlib/install/backend_spec.rb:377
     # ./spec/mixlib/install/backend_spec.rb:50:in `block (3 levels) in <top (required)>'
```

**Pull request merged**